File: system-test/dotnet/SystemTests/Legacy/Mod06/Base/BaseChannelDriverTest.cs
using SystemTests.TestInfrastructure.Configuration;
using Dsl.Core;
using Driver.Port.External.Erp;
using Driver.Adapter.Api;
using Driver.Adapter.Ui;
using Dsl.Core.UseCase;
using Driver.Port;
using Driver.Adapter.External.Erp;
using Driver.Adapter.External.Tax;
using Optivem.Testing;
using Xunit;

namespace SystemTests.Legacy.Mod06.Base;


public abstract class BaseChannelDriverTest : BaseConfigurableTest, IAsyncLifetime
{
    protected IMyShopDriver? _shopDriver;
    protected ErpRealDriver? _erpDriver;
    protected TaxRealDriver? _taxDriver;

    public virtual async Task InitializeAsync()
    {
        await SetupDrivers();
    }

    private async Task SetupDrivers()
    {
        var configuration = LoadConfiguration();

        // Only create shop driver if channel context is set (for channel-parameterized tests)
        // For non-channel tests (like Erp), skip shop driver creation
        try
        {
            _shopDriver = await CreateMyShopDriverAsync(configuration);
        }
        catch (InvalidOperationException ex) when (ex.Message.Contains("Channel type is not set"))
        {
            _shopDriver = null;
        }

        _erpDriver = new ErpRealDriver(configuration.ErpBaseUrl);
        _taxDriver = new TaxRealDriver(configuration.TaxBaseUrl);
    }

    public virtual async Task DisposeAsync()
    {
        if (_shopDriver != null)
            await _shopDriver.DisposeAsync();

        _erpDriver?.Dispose();
        _taxDriver?.Dispose();
    }

    protected async Task SetChannelAsync(Channel channel)
    {
        ChannelContext.Set(channel.Type);
        await SetupDrivers();
    }

    private static async Task<IMyShopDriver?> CreateMyShopDriverAsync(Dsl.Core.Configuration configuration)
    {
        var channelType = ChannelContext.Get();

        if (channelType == ChannelType.UI)
        {
            return await MyShopUiDriver.CreateAsync(configuration.MyShopUiBaseUrl);
        }
        else if (channelType == ChannelType.API)
        {
            return new MyShopApiDriver(configuration.MyShopApiBaseUrl);
        }
        else
        {
            throw new InvalidOperationException($"Unknown channel: {channelType}");
        }
    }
}

File: system-test/java/src/test/java/com/mycompany/myshop/systemtest/legacy/mod06/base/BaseChannelDriverTest.java
package com.mycompany.myshop.systemtest.legacy.mod06.base;

import com.mycompany.myshop.systemtest.configuration.BaseConfigurableTest;
import com.mycompany.myshop.systemtest.configuration.Configuration;
import com.mycompany.myshop.testkit.driver.adapter.external.erp.ErpRealDriver;
import com.mycompany.myshop.testkit.driver.adapter.external.tax.TaxRealDriver;
import com.mycompany.myshop.testkit.channel.ChannelType;
import com.mycompany.myshop.testkit.driver.adapter.api.MyShopApiDriver;
import com.mycompany.myshop.testkit.driver.port.MyShopDriver;
import com.mycompany.myshop.testkit.driver.adapter.ui.MyShopUiDriver;
import com.mycompany.myshop.systemtest.infrastructure.playwright.BrowserLifecycleExtension;
import com.mycompany.myshop.testkit.common.Closer;
import com.optivem.testing.contexts.ChannelContext;
import com.optivem.testing.extensions.ChannelExtension;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.extension.ExtendWith;

@ExtendWith(ChannelExtension.class)
public abstract class BaseChannelDriverTest extends BaseConfigurableTest {
    protected MyShopDriver myShopDriver;
    protected ErpRealDriver erpDriver;
    protected TaxRealDriver taxDriver;

    @BeforeEach
    void setUp() {
        var configuration = loadConfiguration();

        myShopDriver = createChannelMyShopDriver(configuration);
        erpDriver = new ErpRealDriver(configuration.getErpBaseUrl());
        taxDriver = new TaxRealDriver(configuration.getTaxBaseUrl());
    }

    @AfterEach
    void tearDown() {
        Closer.close(myShopDriver);
        Closer.close(erpDriver);
        Closer.close(taxDriver);
    }

    private MyShopDriver createChannelMyShopDriver(Configuration configuration) {
        var channel = ChannelContext.get();

        if(channel == null) {
            return null;
        }

        if (ChannelType.UI.equals(channel)) {
            return new MyShopUiDriver(configuration.getMyShopUiBaseUrl(), BrowserLifecycleExtension.getBrowser());
        } else if (ChannelType.API.equals(channel)) {
            return new MyShopApiDriver(configuration.getMyShopApiBaseUrl());
        } else {
            throw new IllegalStateException("Unknown channel: " + channel);
        }
    }
}
